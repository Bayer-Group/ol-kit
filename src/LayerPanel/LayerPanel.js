import React, { Component } from 'react'
import PropTypes from 'prop-types'
import LayerPanelBase from 'LayerPanel/LayerPanelBase'
import LayerPanelLayersPage from 'LayerPanel/LayerPanelLayersPage'
import { connectToContext } from 'Provider'
/**
 * @component
 * @category LayerPanel
 * @since 0.5.0
 */
class LayerPanel extends Component {
  render () {
    const { translations, children, map, opacity, layerPanelTitle, disableHover, setHoverStyle } = this.props

    return (
      <LayerPanelBase translations={translations} opacity={opacity} layerPanelTitle={layerPanelTitle}>
        {[<LayerPanelLayersPage label='Layers' {...this.props} key='layerPanelLayersPage' map={map} disableHover={disableHover} setHoverStyle={setHoverStyle}/>].concat(children)}
      </LayerPanelBase>
    )
  }
}

LayerPanel.defaultProps = {
  layerPanelTitle: 'Layer Panel',
  disableHover: false
}

LayerPanel.propTypes = {
  /** Object with key/value pairs for translated strings */
  translations: PropTypes.object.isRequired,

  /** An Openlayers map object from which the layer panel will derive its layers */
  map: PropTypes.object.isRequired,

  /** A set of prebuilt page components (like `LayerPanelListPage`) or custom pages */
  children: PropTypes.node,

  /** A float number for the opacity of the LayerPanel. i.e. (0.5) */
  opacity: PropTypes.number,

  /** Title of the LayerPanel */
  layerPanelTitle: PropTypes.string,

  /** Truthy value will disable hover */
  disableHover: PropTypes.bool,

  /** Pass fill, stroke, and color hover style values */
  setHoverStyle: PropTypes.func
}

export default connectToContext(LayerPanel)
