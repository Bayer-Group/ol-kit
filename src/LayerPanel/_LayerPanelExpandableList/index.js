import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ExpandLess from '@material-ui/icons/ExpandMore'
import ExpandMore from '@material-ui/icons/ChevronRight'

class LayerPanelExpandableList extends Component {
  render () {
    const { show, open, handleClick } = this.props

    if (show) {
      if (open) {
        return <ExpandLess data-testid='LayerPanel.collapseLayer' onClick={handleClick} />
      } else {
        return <ExpandMore data-testid='LayerPanel.expandLayer' onClick={handleClick} />
      }
    } else {
      return null
    }
  }
}

LayerPanelExpandableList.propTypes = {
  show: PropTypes.bool,
  open: PropTypes.bool,
  handleClick: PropTypes.func.isRequired
}

export default LayerPanelExpandableList
