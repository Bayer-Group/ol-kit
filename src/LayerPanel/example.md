The `LayerPanel` is a convenient way to view things that are on the map. The original intent is to view layers but with this `LayerPanel` you can view whatever you want. Build pages into the `LayerPanel` that will show up in their own tabs to make whatever customization you want. Another common page that works well with the `LayerPanel` is a Legends page which you can make with a `LayerPanelPage` component.

Its as simple as that. This will give you a `LayerPanel` placed on the right side of the screen that allows you to view the Layers from the map and toggle visibility/remove.
```jsx
import { LayerPanel } from '@bayer/ol-kit'

return <LayerPanel />
```
![image](https://github.platforms.engineering/storage/user/2990/files/cb487900-86d2-11ea-8a6c-587e8bb8377c)

We also have a prebuilt page that ships with the `LayerPanel` called `LayerPanelLayersPage`. It's a simple drop in that allows you to customize things like handleFeatureDoubleClick of a list item and quite a few other props. Check out the docs on the `LayerPanelLayersPage` to see more of the props you can pass. You can either customize it by importing it yourself (which requires you to switch to the `LayerPanelBase` as the parent component) or you can pass the props of the `LayerPanelLayersPage` down through the prebuilt `LayerPanel`.
```jsx
import { LayerPanel, LayerPanelPage, LayerPanelHeader, LayerPanelContent } from '@bayer/ol-kit'
import VpnKeyIcon from '@material-ui/icons/VpnKeyIcon'

// Passing the props through the LayerPanel component
return (
  <LayerPanel handleFeatureDoubleClick={this.onFeatureDoubleClick} layerPanelTitle='My Layer Panel'>
    <LayerPanelPage tabIcon={<VpnKeyIcon />}>
      <LayerPanelHeader
        title='Legends Page'
        avatar={<VpnKeyIcon />} />
      <LayerPanelContent>
        Hello world
      </LayerPanelContent>
    </LayerPanelPage>
  </LayerPanel>
)

// Directly pulling in the LayerPanelLayersPage and passing props through it
return (
  <LayerPanelBase>
    <LayerPanelLayersPage handleFeatureDoubleClick={this.onFeatureDoubleClick}>
    <LayerPanelPage tabIcon={<VpnKeyIcon />}>
      <LayerPanelHeader
        title='Legends Page'
        avatar={<VpnKeyIcon />} />
      <LayerPanelContent>
        Hello world
      </LayerPanelContent>
    </LayerPanelPage>
  </LayerPanelBase>
)
```

To show you the amount of customization possible with these pages, here's the `LayerPanelLayersPage`.
```jsx
import {
  LayerPanel,
  LayerPanelPage,
  LayerPanelHeader,
  LayerPanelContent,
  LayerPanelActions,
  LayerPanelCheckbox,
  LayerPanelActionRemove,
  LayerPanelActionImport,
  LayerPanelActionExport,
  LayerPanelList,
  LayerPanelListItem } from '@bayer/ol-kit'
import VpnKeyIcon from '@material-ui/icons/VpnKeyIcon'
import MoreHorizIcon from '@material-ui/icons/MoreHorizIcon'

return (
  <LayerPanelPage>
    <LayerPanelHeader
      title={'Active Layers'}
      translations={translations}
      avatar={<LayerPanelCheckbox checkboxState={masterCheckboxVisibility} handleClick={this.setVisibilityForAllLayers} />}
      actions={customActions ||
        <LayerPanelActions
          icon={<MoreHorizIcon />}
          translations={translations}
          layers={layers}
          map={map}>
          <LayerPanelActionRemove
            removeFeaturesForLayer={this.removeFeaturesForLayer}
            shouldAllowLayerRemoval={shouldAllowLayerRemoval} />
          {onFileImport && <LayerPanelActionImport handleImport={onFileImport} />}
          {onExportFeatures && <LayerPanelActionExport onExportFeatures={onExportFeatures} />}
        </LayerPanelActions>} />
    {enableFilter &&
      <TextField
        id='feature-filter-input'
        label={'Filter Layers'}
        type='text'
        style={{ margin: '8px' }}
        fullWidth
        value={filterText}
        onChange={(e) => this.handleFilter(e.target.value)} />
    }
    <LayerPanelContent>
      <LayerPanelList disableDrag={disableDrag} onSort={this.zIndexSort} onReorderedItems={this.reorderLayers} items={layers} >
        {layerFilter(layers).filter((layer) => {
          const filteredFeatures = this.getFeaturesForLayer(layer)

          return !enableFilter || !(layer instanceof olLayerVector) ? true : filteredFeatures.length
        }).sort(this.zIndexSort).map((layer, i) => {
          const features = this.getFeaturesForLayer(layer)

          return (
            <div key={i}>
              <LayerPanelListItem handleDoubleClick={() => { handleLayerDoubleClick(layer) }}>
                {<LayerPanelCheckbox
                  checkboxState={!layer ? null : layer.getVisible()}
                  handleClick={(e) => this.handleVisibility(e, layer)} />}
                {<LayerPanelExpandableList show={!!features} open={expandedLayer} handleClick={this.handleExpandedLayer} />}
                <ListItemText primary={layer.get('title') || 'Untitled Layer'} />
                <ListItemSecondaryAction>
                  <LayerPanelActions
                    icon={<MoreVertIcon />}
                    translations={translations}
                    layer={layer}
                    map={map} >
                    {getMenuItemsForLayer(layer) ||
                    [<LayerPanelActionRemove key='removeLayer' shouldAllowLayerRemoval={shouldAllowLayerRemoval} />,
                      <LayerPanelActionExtent key='gotoExtent' />,
                      <LayerPanelActionOpacity key='layerOpacity' />]}
                  </LayerPanelActions>
                </ListItemSecondaryAction>
              </LayerPanelListItem>
              { features
                ? <Collapse in={expandedLayer} timeout='auto' unmountOnExit>
                  <List component='div' disablePadding style={{ paddingLeft: '36px' }}>
                    {features.map((feature, i) => {
                      return (
                        <ListItem key={i} hanldeDoubleClick={() => handleFeatureDoubleClick(feature)}>
                          <LayerPanelCheckbox
                            handleClick={(event) => this.handleFeatureCheckbox(layer, feature, event)}
                            checkboxState={feature.get('_feature_visibility')} />
                          <ListItemText inset={false} primary={feature.get('name') || `Feature ${i}`} />
                        </ListItem>
                      )
                    })}
                  </List>
                </Collapse> : null }
            </div>
          )
        })}
      </LayerPanelList>
    </LayerPanelContent>
  </LayerPanelPage>
)
```










