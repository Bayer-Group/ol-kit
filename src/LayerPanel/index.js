export { default as LayerPanel } from './LayerPanel'
export { default as LayerPanelBase } from './LayerPanelBase'
export { default as LayerPanelActions } from './LayerPanelActions'
export { default as LayerPanelCheckbox } from './LayerPanelCheckbox'
export { default as LayerPanelContent } from './LayerPanelContent'
export { default as LayerPanelHeader } from './LayerPanelHeader'
export { default as LayerPanelList } from './LayerPanelList'
export { default as LayerPanelListItem } from './LayerPanelListItem'
export { default as LayerPanelLayersPage } from './LayerPanelLayersPage'
export { default as LayerPanelMenu } from './LayerPanelMenu'
export { default as LayerPanelPage } from './LayerPanelPage'
export { default as LayerPanelActionRemove } from './LayerPanelActionRemove'
export { default as LayerPanelActionExport } from './LayerPanelActionExport'
export { default as LayerPanelActionExtent } from './LayerPanelActionExtent'
export { default as LayerPanelActionImport } from './LayerPanelActionImport'
export { default as LayerPanelActionOpacity } from './LayerPanelActionOpacity'
export * from './LayerPanelActionExport/utils'
