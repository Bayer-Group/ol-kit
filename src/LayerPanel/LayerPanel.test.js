import React from 'react'
import { render, waitFor, fireEvent } from '@testing-library/react'
import olFeature from 'ol/Feature'
import olLayerVector from 'ol/layer/Vector'
import olPoint from 'ol/geom/Point'
import olSourceVector from 'ol/source/Vector'
import olFormatGeoJSON from 'ol/format/GeoJSON'
import { Map } from 'Map'
import { LayerPanel } from 'LayerPanel'
const fs = require('fs')

const testGeoJSONData = {
  type: 'FeatureCollection',
  crs: {
    type: 'name',
    properties: {
      name: 'EPSG:3857'
    }
  },
  features: [{
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [0, 0]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'MultiPoint',
      coordinates: [[-2e6, 0], [0, -2e6]]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'LineString',
      coordinates: [[4e6, -2e6], [8e6, 2e6], [9e6, 2e6]]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'LineString',
      coordinates: [[4e6, -2e6], [8e6, 2e6], [8e6, 3e6]]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'Polygon',
      coordinates: [[[-5e6, -1e6], [-4e6, 1e6],
        [-3e6, -1e6], [-5e6, -1e6]], [[-4.5e6, -0.5e6],
        [-3.5e6, -0.5e6], [-4e6, 0.5e6], [-4.5e6, -0.5e6]]]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'MultiLineString',
      coordinates: [
        [[-1e6, -7.5e5], [-1e6, 7.5e5]],
        [[-1e6, -7.5e5], [-1e6, 7.5e5], [-5e5, 0], [-1e6, -7.5e5]],
        [[1e6, -7.5e5], [15e5, 0], [15e5, 0], [1e6, 7.5e5]],
        [[-7.5e5, -1e6], [7.5e5, -1e6]],
        [[-7.5e5, 1e6], [7.5e5, 1e6]]
      ]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'MultiPolygon',
      coordinates: [
        [[[-5e6, 6e6], [-5e6, 8e6], [-3e6, 8e6],
          [-3e6, 6e6], [-5e6, 6e6]]],
        [[[-3e6, 6e6], [-2e6, 8e6], [0, 8e6],
          [0, 6e6], [-3e6, 6e6]]],
        [[[1e6, 6e6], [1e6, 8e6], [3e6, 8e6],
          [3e6, 6e6], [1e6, 6e6]]]
      ]
    }
  }, {
    type: 'Feature',
    geometry: {
      type: 'GeometryCollection',
      geometries: [{
        type: 'LineString',
        coordinates: [[-5e6, -5e6], [0, -5e6]]
      }, {
        type: 'Point',
        coordinates: [4e6, -5e6]
      }, {
        type: 'Polygon',
        coordinates: [
          [[1e6, -6e6], [2e6, -4e6], [3e6, -6e6], [1e6, -6e6]]
        ]
      }]
    }
  }]
}

describe('<LayerPanel />', () => {
  it('should filter out basemap', async () => {
    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })

    const { getByText, getByTestId } = render(<Map onMapInit={onMapInit}><LayerPanel /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    fireEvent.click(getByTestId('LayerPanel.open'))

    // add a feature to that map at a known pixel location
    const features = [new olFeature(new olPoint([-97.75, 30.265]))]
    const source = new olSourceVector({ features })
    const vectorLayer = new olLayerVector({ source })

    vectorLayer.set('title', 'My Custom Layer')
    testMap.addLayer(vectorLayer)
    const basemap = testMap.getLayers().getArray().find(layer => layer.get('_ol_kit_basemap'))

    basemap.set('title', 'My Basemap')

    expect(testMap.getLayers().getArray().length).toBe(2)

    await waitFor(() => expect(getByText('My Custom Layer')).toBeInTheDocument())
  })
  it('should toggle visibility on checkbox click', async () => {
    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })

    const { getByTestId, getAllByTestId } = render(<Map onMapInit={onMapInit}><LayerPanel /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    fireEvent.click(getByTestId('LayerPanel.open'))

    // add a feature to that map at a known pixel location
    const features = [new olFeature(new olPoint([-97.75, 30.265]))]

    const source = new olSourceVector({ features })
    const vectorLayer = new olLayerVector({ source })
    const vectorLayer2 = new olLayerVector({ source })

    vectorLayer.set('title', 'My Custom Layer')
    vectorLayer2.set('title', 'My Second Custom Layer')
    testMap.addLayer(vectorLayer2)
    testMap.addLayer(vectorLayer)

    const checkboxes = getAllByTestId('LayerPanel.checked')

    fireEvent.click(checkboxes[0])

    // make sure the basemap is still visible and that the only layer rendered in the layer panel is not visible
    expect(testMap.getLayers().getArray()[0].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[1].getVisible()).toBe(false)
    expect(testMap.getLayers().getArray()[2].getVisible()).toBe(false)

    fireEvent.click(checkboxes[0])

    // make sure the basemap is still visible and that the only layer rendered in the layer panel is not visible
    expect(testMap.getLayers().getArray()[0].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[1].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[2].getVisible()).toBe(true)

    fireEvent.click(checkboxes[1])

    // when we toggle visibility of one but not all layers the master checkbox should be indeterminate
    expect(testMap.getLayers().getArray()[0].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[1].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[2].getVisible()).toBe(false)
    expect(getByTestId('LayerPanel.indeterminateCheckbox')).toBeInTheDocument()

    fireEvent.click(checkboxes[0])
    expect(testMap.getLayers().getArray()[0].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[1].getVisible()).toBe(true)
    expect(testMap.getLayers().getArray()[2].getVisible()).toBe(true)
  })
  it('should remove all Layers from action bar', async () => {
    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })

    const { getByTestId, getAllByTestId, getByText } = render(<Map onMapInit={onMapInit}><LayerPanel /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    fireEvent.click(getByTestId('LayerPanel.open'))

    // add a feature to that map at a known pixel location
    const features = [new olFeature(new olPoint([-97.75, 30.265]))]

    const source = new olSourceVector({ features })
    const vectorLayer = new olLayerVector({ source })
    const vectorLayer2 = new olLayerVector({ source })

    vectorLayer.set('title', 'My Custom Layer')
    vectorLayer2.set('title', 'My Second Custom Layer')
    testMap.addLayer(vectorLayer2)
    testMap.addLayer(vectorLayer)

    const actions = getAllByTestId('LayerPanel.actionsButton')

    fireEvent.click(actions[0])
    await waitFor(() => expect(getByText('Remove All Layers')).toBeInTheDocument())

    fireEvent.click(getByText('Remove All Layers'))

    // should still have the basemap
    expect(testMap.getLayers().getArray().length).toBe(1)

    testMap.addLayer(vectorLayer2)
    testMap.addLayer(vectorLayer)

    // basemap is on the map still to +1
    expect(testMap.getLayers().getArray().length).toBe(3)

    const checkboxes = getAllByTestId('LayerPanel.unchecked')

    fireEvent.click(checkboxes[1])

    // make sure the layer that was toggled off is not removed
    fireEvent.click(actions[0])
    await waitFor(() => expect(getByText('Remove All Layers')).toBeInTheDocument())

    fireEvent.click(getByText('Remove All Layers'))

    // should still have the basemap
    expect(testMap.getLayers().getArray().length).toBe(2)
    await waitFor(() => expect(getByText('My Second Custom Layer')).toBeInTheDocument())
  })

  it('should render features', async () => {
    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })

    const { getByTestId, getAllByTestId, getByText } = render(<Map onMapInit={onMapInit}><LayerPanel /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    fireEvent.click(getByTestId('LayerPanel.open'))

    // add a feature to that map at a known pixel location
    const features = [new olFeature(new olPoint([-97.75, 30.265])), new olFeature(new olPoint([-99.75, 35.265]))]

    features[0].set('name', 'My map feature')

    const source = new olSourceVector({ features })
    const vectorLayer = new olLayerVector({ source })

    vectorLayer.set('title', 'My Custom Layer')
    testMap.addLayer(vectorLayer)

    fireEvent.click(getByTestId('LayerPanel.expandLayer'))

    await waitFor(() => expect(getByText('My map feature')).toBeInTheDocument())

    const checkboxes = getAllByTestId('LayerPanel.checked')

    fireEvent.click(checkboxes[2])

    // when we toggle visibility of one but not all features the master checkbox and the layers checkbox should be indeterminate
    expect(getAllByTestId('LayerPanel.indeterminateCheckbox').length).toBe(2)
  })

  it('should export GeoJSON data', async () => {
    const mockCreateFn = global.URL.createObjectURL = jest.fn()
    const mockRevokeFn = global.URL.revokeObjectURL = jest.fn()

    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })

    const { getByTestId, getAllByTestId } = render(<Map onMapInit={onMapInit}><LayerPanel /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    fireEvent.click(getByTestId('LayerPanel.open'))

    const vectorSource = new olSourceVector({
      features: (new olFormatGeoJSON()).readFeatures(testGeoJSONData)
    })
    const vectorLayer = new olLayerVector({
      title: 'TestLayer',
      mode: 'image',
      source: vectorSource
    })

    testMap.addLayer(vectorLayer)

    const actions = getAllByTestId('LayerPanel.actionsButton')

    fireEvent.click(actions[0])
    fireEvent.click(getByTestId('LayerPanel.exportGeoJSON'))

    const download = fs.readFileSync('export.geojson', { encoding: 'utf8' })

    expect(download).toMatchSnapshot()
    fs.unlink('./export.geojson', jest.fn())
    await waitFor(() => expect(mockCreateFn).toHaveBeenCalled()) // creates an object url
    // await waitFor(() => expect(getByTestId('Export.download')).toBeInTheDocument()) // expects the <a> node to be appended
    await waitFor(() => expect(mockRevokeFn).toHaveBeenCalled()) // cleans up the object url
  })

  it('should export KML data', async () => {
    const mockCreateFn = global.URL.createObjectURL = jest.fn()
    const mockRevokeFn = global.URL.revokeObjectURL = jest.fn()

    let testMap
    const onMapInit = jest.fn(map => {
      testMap = map
    })

    const { getByTestId, getAllByTestId } = render(<Map onMapInit={onMapInit}><LayerPanel /></Map>)

    // wait for async child render
    await waitFor(() => expect(onMapInit).toHaveBeenCalled())

    fireEvent.click(getByTestId('LayerPanel.open'))

    const vectorSource = new olSourceVector({
      features: (new olFormatGeoJSON()).readFeatures(testGeoJSONData)
    })
    const vectorLayer = new olLayerVector({
      title: 'TestLayer',
      mode: 'image',
      source: vectorSource
    })

    testMap.addLayer(vectorLayer)

    const actions = getAllByTestId('LayerPanel.actionsButton')

    fireEvent.click(actions[0])
    fireEvent.click(getByTestId('LayerPanel.exportKML'))

    const download = fs.readFileSync('export.kml', { encoding: 'utf8' })

    expect(download).toMatchSnapshot()
    fs.unlink('export.kml', jest.fn())
    await waitFor(() => expect(mockCreateFn).toHaveBeenCalled()) // creates an object url
    // await waitFor(() => expect(getByTestId('Export.download')).toBeInTheDocument()) // expects the <a> node to be appended
    await waitFor(() => expect(mockRevokeFn).toHaveBeenCalled()) // cleans up the object url
  })
})
