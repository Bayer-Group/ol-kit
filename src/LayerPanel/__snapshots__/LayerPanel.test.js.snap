// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LayerPanel /> should export GeoJSON data 1`] = `"{\\"type\\":\\"FeatureCollection\\",\\"features\\":[{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"Point\\",\\"coordinates\\":[0,0]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiPoint\\",\\"coordinates\\":[[-17.966305682390427,0],[0,-17.678914238335736]]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[35.932611364780854,-17.678914238335736],[71.86522272956171,17.67891423833575],[80.84837557075693,17.67891423833575]]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"LineString\\",\\"coordinates\\":[[35.932611364780854,-17.678914238335736],[71.86522272956171,17.67891423833575],[71.86522272956171,26.007420231981243]]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"Polygon\\",\\"coordinates\\":[[[-44.91576420597607,-8.946573850543416],[-35.932611364780854,8.946573850543416],[-26.94945852358564,-8.946573850543416],[-44.91576420597607,-8.946573850543416]],[[-40.424187785378464,-4.486983030705048],[-31.441034944183247,-4.486983030705048],[-35.932611364780854,4.486983030705062],[-40.424187785378464,-4.486983030705048]]]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiLineString\\",\\"coordinates\\":[[[-8.983152841195214,-6.721891585698515],[-8.983152841195214,6.721891585698515]],[[-8.983152841195214,-6.721891585698515],[-8.983152841195214,6.721891585698515],[-4.491576420597607,0],[-8.983152841195214,-6.721891585698515]],[[8.983152841195214,-6.721891585698515],[13.47472926179282,0],[13.47472926179282,0],[8.983152841195214,6.721891585698515]],[[-6.73736463089641,-8.946573850543416],[6.73736463089641,-8.946573850543416]],[[-6.73736463089641,8.946573850543416],[6.73736463089641,8.946573850543416]]]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"MultiPolygon\\",\\"coordinates\\":[[[[-44.91576420597607,47.353704702433134],[-44.91576420597607,58.15524209465974],[-26.94945852358564,58.15524209465974],[-26.94945852358564,47.353704702433134],[-44.91576420597607,47.353704702433134]]],[[[-26.94945852358564,47.353704702433134],[-17.966305682390427,58.15524209465974],[0,58.15524209465974],[0,47.353704702433134],[-26.94945852358564,47.353704702433134]]],[[[8.983152841195214,47.353704702433134],[8.983152841195214,58.15524209465974],[26.94945852358564,58.15524209465974],[26.94945852358564,47.353704702433134],[8.983152841195214,47.353704702433134]]]]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}},{\\"type\\":\\"Feature\\",\\"geometry\\":{\\"type\\":\\"GeometryCollection\\",\\"geometries\\":[{\\"type\\":\\"LineString\\",\\"coordinates\\":[[-44.91576420597607,-40.91627447067578],[0,-40.91627447067578]]},{\\"type\\":\\"Point\\",\\"coordinates\\":[35.932611364780854,-40.91627447067578]},{\\"type\\":\\"Polygon\\",\\"coordinates\\":[[[8.983152841195214,-47.353704702433106],[17.966305682390427,-33.78523007002315],[26.94945852358564,-47.353704702433106],[8.983152841195214,-47.353704702433106]]]}]},\\"properties\\":{\\"_ol_kit_feature_visibility\\":true,\\"_ol_kit_feature_style\\":null,\\"_ol_kit_parent\\":null}}]}"`;

exports[`<LayerPanel /> should export KML data 1`] = `"<kml xmlns=\\"http://www.opengis.net/kml/2.2\\" xmlns:gx=\\"http://www.google.com/kml/ext/2.2\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://www.opengis.net/kml/2.2 https://developers.google.com/kml/schema/kml22gx.xsd\\"><Document><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><Point><coordinates>0,0</coordinates></Point></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><MultiGeometry><Point><coordinates>-17.966305682390427,0</coordinates></Point><Point><coordinates>0,-17.678914238335736</coordinates></Point></MultiGeometry></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><LineString><coordinates>35.932611364780854,-17.678914238335736 71.86522272956171,17.67891423833575 80.84837557075693,17.67891423833575</coordinates></LineString></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><LineString><coordinates>35.932611364780854,-17.678914238335736 71.86522272956171,17.67891423833575 71.86522272956171,26.007420231981243</coordinates></LineString></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><Polygon><innerBoundaryIs><LinearRing><coordinates>-40.424187785378464,-4.486983030705048 -31.441034944183247,-4.486983030705048 -35.932611364780854,4.486983030705062 -40.424187785378464,-4.486983030705048</coordinates></LinearRing></innerBoundaryIs><outerBoundaryIs><LinearRing><coordinates>-44.91576420597607,-8.946573850543416 -35.932611364780854,8.946573850543416 -26.94945852358564,-8.946573850543416 -44.91576420597607,-8.946573850543416</coordinates></LinearRing></outerBoundaryIs></Polygon></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><MultiGeometry><LineString><coordinates>-8.983152841195214,-6.721891585698515 -8.983152841195214,6.721891585698515</coordinates></LineString><LineString><coordinates>-8.983152841195214,-6.721891585698515 -8.983152841195214,6.721891585698515 -4.491576420597607,0 -8.983152841195214,-6.721891585698515</coordinates></LineString><LineString><coordinates>8.983152841195214,-6.721891585698515 13.47472926179282,0 13.47472926179282,0 8.983152841195214,6.721891585698515</coordinates></LineString><LineString><coordinates>-6.73736463089641,-8.946573850543416 6.73736463089641,-8.946573850543416</coordinates></LineString><LineString><coordinates>-6.73736463089641,8.946573850543416 6.73736463089641,8.946573850543416</coordinates></LineString></MultiGeometry></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><MultiGeometry><Polygon><outerBoundaryIs><LinearRing><coordinates>-44.91576420597607,47.353704702433134 -44.91576420597607,58.15524209465974 -26.94945852358564,58.15524209465974 -26.94945852358564,47.353704702433134 -44.91576420597607,47.353704702433134</coordinates></LinearRing></outerBoundaryIs></Polygon><Polygon><outerBoundaryIs><LinearRing><coordinates>-26.94945852358564,47.353704702433134 -17.966305682390427,58.15524209465974 0,58.15524209465974 0,47.353704702433134 -26.94945852358564,47.353704702433134</coordinates></LinearRing></outerBoundaryIs></Polygon><Polygon><outerBoundaryIs><LinearRing><coordinates>8.983152841195214,47.353704702433134 8.983152841195214,58.15524209465974 26.94945852358564,58.15524209465974 26.94945852358564,47.353704702433134 8.983152841195214,47.353704702433134</coordinates></LinearRing></outerBoundaryIs></Polygon></MultiGeometry></Placemark><Placemark><ExtendedData><Data name=\\"_ol_kit_feature_style\\"/><Data name=\\"_ol_kit_feature_visibility\\"><value>true</value></Data><Data name=\\"_ol_kit_parent\\"/></ExtendedData><MultiGeometry><LineString><coordinates>-44.91576420597607,-40.91627447067578 0,-40.91627447067578</coordinates></LineString><Point><coordinates>35.932611364780854,-40.91627447067578</coordinates></Point><Polygon><outerBoundaryIs><LinearRing><coordinates>8.983152841195214,-47.353704702433106 17.966305682390427,-33.78523007002315 26.94945852358564,-47.353704702433106 8.983152841195214,-47.353704702433106</coordinates></LinearRing></outerBoundaryIs></Polygon></MultiGeometry></Placemark></Document></kml>"`;
